man/syrep.1
.TH syrep 1 User Manuals
.SH NAME
syrep \- A file repository synchronization tool
.SH SYNOPSIS
\fBsyrep [options...\fB] --list\fB \fISNAPSHOT\fB ...

syrep [options...\fB] --info\fB \fISNAPSHOT\fB ...

syrep [options...\fB] --history\fB \fISNAPSHOT\fB ...

syrep [options...\fB] --dump\fB \fISNAPSHOT\fB ...

syrep [options...\fB] --update\fB \fIDIRECTORY\fB ...

syrep [options...\fB] --diff\fB \fISNAPSHOT\fB \fISNAPSHOT\fB

syrep [options...\fB] --merge\fB \fISNAPSHOT\fB \fIDIRECTORY\fB

syrep [options...\fB] --merge\fB \fIPATCH\fB \fIDIRECTORY\fB

syrep [options...\fB] --merge\fB \fIDIRECTORY\fB \fIDIRECTORY\fB

syrep [options...\fB] --makepatch\fB \fIDIRECTORY\fB \fISNAPSHOT\fB

syrep [options...\fB] --extract\fB \fISNAPSHOT\fB ...

syrep [options...\fB] --cleanup\fB \fIDIRECTORY\fB ...

syrep [options...\fB] --forget\fB \fISNAPSHOT\fB ...
\f1
.SH DESCRIPTION
Syrep is a generic file repository synchronization tool. It may be used to synchronize large file hierarchies bidirectionally by exchanging patch files. Syrep is truly peer-to-peer, no central servers are involved. Synchronizations between more than two repositories are supported. The patch files may be transferred via offline media, e.g. removable hard disks or compact discs.

Files are tracked by their message digests, currently MD5.

Syrep was written to facilitate the synchronization of two large digital music repositories without direct network connection. Patch files of several gigabytes are common in this situation.

Syrep is able to cope with 64 bit file sizes. (LFS)

Syrep is optimized for speed. It may make use of a message digest cache to accelerate the calculation of digests of a whole directory hierarchy.

A syrep repository is a normal UNIX directory tree containing a special directory \fI.syrep\f1 with a file \fIcurrent.syrep\f1 (called snapshot) which holds file system history data. A directory is turned into a syrep repository by running \fB--update\f1 on it. Snapshots are used to perform basic tasks like comparing repositories (command \fB--diff\f1) or creating patches between them (command \fB--make-patch\f1).

Syrep will ignore all files and directories that have the extended attribute \fBuser.syrep\f1 set to \fBomit\f1. Currently, this works on Linux only. Keep in mind, that only newer kernel versions and some file system support extended attributes. See \fBattr(5)\f1 for more information.
.SH GENERAL OPTIONS
Exactly one command has to be specified on the command line. On the other hand multiple options are allowed.
.TP
\fB-v\f1 | \fB--verbose\f1 (option)
Enable more verbose operation 
.TP
\fB-T\f1 | \fB--local-temp\f1 (option)
Use temporary directory inside repository. This is very useful when the file repository you apply patches to is on a different partition than /tmp, because files my be hard linked instead of copied. This requires a read-writable file system however. 
.TP
\fB--ignore-origin\f1 (option)
Normally syrep warns you if you update, merge or makepatch a repository with a matching snapshot not generated on the original host, and asks the user if he really wants to proceed. This option may be used to disable this question. 
.TP
\fB-z\f1 | \fB--compress\f1 (option)
Compress output snapshots or patches. This may slow down syrep extraordinarily and is more or less useless if the data to compress is already compressed. I suggest using it for --update but not for --makepatch if the file repository contains MP3 or MPEG files only. 
.TP
\fB-p\f1 | \fB--progress\f1 (option)
Show a rotating dash while executing operations 
.SH GENERAL COMMANDS
.TP
\fB-h\f1 | \fB--help\f1 (command)
Print help and exit 
.TP
\fB-V\f1 | \fB--version\f1 (command)
Print version information and exit 
.SH LISTING SNAPSHOT CONTENTS
.TP
\fB--list\f1 (command)
Command for listing the file log of a repository snapshot 
.TP
\fB--show-deleted\f1 (option)
Show deleted entries 
.TP
\fB--show-by-md\f1 (option)
Show files by message digests. This option collides with \fB--sort\f1. 
.TP
\fB--show-times\f1 (option)
Show first and last seen times 
.TP
\fB--sort\f1 (option)
Sort file listing chronologically. This option collides with \fB--show-by-md\f1. 
.SH SHOWING SNAPSHOT INFORMATION
.TP
\fB--info\f1 (command)
Show information about a repository or snapshot, such as origin, current timestamp and version, database size. 
.SH SHOWING SNAPSHOT HISTORY
.TP
\fB--history\f1 (command)
Show the version and timestamp history of a snapshot's updates 
.SH DUMPING SNAPSHOT CONTENTS
.TP
\fB--dump\f1 (command)
Show a structure dump of a repository or snapshot 
.SH UPDATING A SNAPSHOT
.TP
\fB--update\f1 (command)
Update or create a snapshot for a repository directory. That is: iterate through the specified hierarchy and update the snapshot log information accordingly. 
.TP
\fB-S\f1\fISTRING\f1 | \fB--snapshot=\f1\fISTRING\f1 (option)
Use the specified snapshot file instead of the one contained in the repository directory. This may be helpful if your file hierarchy is read only. 
.TP
\fB-C\f1\fISTRING\f1 | \fB--cache=\f1\fISTRING\f1 (option)
Use the specified message digest cache file instead of the one contained in the repository directory. This may be helpful if your file hierarchy is read only or when you plan to maintain a system wide message digest cache. In the latter case you should use \fB--no-purge\f1 as well. 
.TP
\fB--no-cache\f1 (option)
Don't use a message digest cache. 
.TP
\fB--no-purge\f1 (option)
Don't purge obsolete entries from cache after update run. The may be useful if you plan to maintain a system wide message digest cache. 
.TP
\fB--ro-cache\f1 (option)
Use cache in a read only fashion 
.TP
\fB--check-dev\f1 (option)
Store information about the device where the file resides when storing an entry about it in the message digest cache. Since nowadays device identifiers cannot be longer considered stable, this options defaults to off. 
.SH SHOWING DIFFERENCES BETWEEN TWO SNAPSHOTS
.TP
\fB--diff\f1 (command)
Show difference between two repositories or snapshots 
.TP
\fB-s\f1 | \fB--sizes\f1 (option)
Show the sizes of the files to copy. This works only when acting on a local repository and for the local files. 
.TP
\fB-H\f1 | \fB--human-readable\f1 (option)
Only useful when using with \fB-s\f1. When enabled shows file sizes in human readable form, i.e. "3.5 MB" instead of "3670016". 
.SH MERGING A SNAPSHOT, A PATCH OR A REPOSITORY INTO A REPOSITORY
.TP
\fB--merge\f1 (command)
Merge a snapshot, a patch or a repository into a repository. Afterwards, you should run \fB--update\f1 on the repository to update the snapshot. 
.TP
\fB-q\f1 | \fB--question\f1 (option)
Ask a question before each action 
.TP
\fB-P\f1 | \fB--prune-empty\f1 (option)
Prune empty directories 
.TP
\fB--keep-trash\f1 (option)
Don't empty trash. Deleted files are copied into a trash folder inside the repository directory. If this option is specified this trash is not emptied when the operation is completed. 
.TP
\fB--check-md\f1 (option)
Check message digests of files before deleting or replacing them. NB: This worsens syrep's performance and is thus not enabled by default. 
.TP
\fB--always-copy\f1 (option)
As default syrep tries to hard link files instead of copying them. With this option syrep will always do a copy, regardless if a hard link is possible or not. This is especially useful when doing bi-directory merges. NB: This worsens syrep's performance and is thus not enabled by default. 
.SH MAKING A PATCH FOR A SNAPSHOT AGAINST A DIRECTORY
.TP
\fB--makepatch\f1 (command)
Make a patch against the specified repository. The patch is written to STDOUT unless \fB-o\f1 is specified. 
.TP
\fB-o\f1\fISTRING\f1 | \fB--output-file=\f1\fISTRING\f1 (option)
Write output to specified file instead of STDOUT 
.TP
\fB--include-all\f1 (option)
Include files in patch which do exist on the other side under a different name 
.SH EXTRACTING A SNAPSHOT'S CONTENTS
.TP
\fB--extract\f1 (command)
Extract the contents of a snapshot or patch to the local directory unless \fB-D\f1 is specified. 
.TP
\fB-D\f1\fISTRING\f1 | \fB--output-directory=\f1\fISTRING\f1 (option)
Write output to specified directory 
.SH CLEANING UP A REPOSITORY
.TP
\fB--cleanup\f1 (command)
Remove syrep info from repository 
.TP
\fB-l\f1\fIINT\f1 | \fB--cleanup-level=\f1\fIINT\f1
1: just remove temporary data and trash (default); 2: remove message digest cache as well; 3: remove all syrep data 
.SH FORGETTING OUTDATED SNAPSHOT INFORMATION
.TP
\fB--forget\f1 (command)
Repackage the snapshot file dropping outdated information. Use this if your snapshot files get too large. 
.TP
\fB-R\f1 \fIDAYS\f1 | \fB--remember=\f1\fIDAYS\f1 (option)
Information about how many days in the past should be kept? This defaults to 180 (half a year). 
.SH REPOSITORIES, SNAPSHOTS AND PATCHES
A syrep file repository is a POSIX file hierarchy with some additional log data, which is used to track changes. Normally this log data is saved as "snapshot" in the file \fI$(REPOSITORY)/.syrep/current.syrep\f1. You may create and update it by running \fB--update\f1. The more often this log is updated the better modifications may be tracked. Therefore this operation should be called at least once a day via \fBcron(8)\f1

Two snapshots of two distinct repositories (possibly from different hosts) may be compared with \fB---diff\f1. This will show you which files should be copied or deleted from or to the other repository. \fB--makepatch\f1 will attach the data of the local missing in the remote repository to a snapshot and write it to a patch file. This file should be transferred to the other repository and applied there with \fB--merge\f1.

Keep in mind that patches contain the snapshot data of the originating host. Because of that you may use it as a snapshot, e.g. by running \fB--diff\f1 on it. On the other hand you are also able to merge snapshots without attached patch data to a repository. This will do all required deletions and renames, but naturally won't add any new data to the file tree.

To extract the contents of a patch you may use \fB--extract\f1. This will write all files contained in the patch or snapshot to the local directory, including snapshot log data. Files are named by their message digests.
.SH FILES
\fI$(REPOSITORY)/.syrep/current.syrep\f1 is the current snapshot of the repository. It is created and updated by running \fB--update\f1 on the directory. Use this file to create patches on other repositories against this one. This file may be compressed by specifying \fB--compress\f1 when running \fB--update\f1.

\fI$(REPOSITORY)/.syrep/md-cache\f1 is the message digest cache which may be used to accelerate the repeated operation of \fB---update\f1. It associates device numbers, inode numbers, file sizes and modification times with the message digest calculated for that file. The file is only valid on the host it was created on since it contains device numbers. 

\fI$(REPOSITORY)/.syrep/trash/\f1 is the trash directory used by \fB--merge\f1. Files are moved in here on deletion. After successful completion it is emptied unless \fB--keep-trash\f1 is specified.

\fI$(REPOSITORY)/.syrep/tmp/\f1 is used as temporary file space for extracting snaphots when option \fB--local-temp\f1 is used.
.SH RETURN VALUES
\fI0\f1 Success

\fINonzero\f1 Failure
.SH AUTHOR
Syrep was written by Lennart Poettering <mzflerc (at) 0pointer (dot) de>.

Syrep is available at \fBhttp://0pointer.de/lennart/projects/syrep/\f1

You are encouraged to improve this documentation, feel free to send me patches. This is free software, after all.
.SH SEE ALSO
\fBrsync(1)\f1, \fBcron(8)\f1, \fBattr(5)\f1
.SH COMMENTS
This man page was written using \fBxml2man(1)\f1 by Oliver Kurth.


debian/changelog
syrep (0.8-1) unstable; urgency=low

  * Upstream: update origin field of snapshots on update

 -- Lennart Poettering <mzflerc@0pointer.net>  Sun, 16 Apr 2006 20:24:59 +0200

syrep (0.7-2) unstable; urgency=low

  * Fix README link
  * Fix dependency on libattr1

 -- Lennart Poettering <mzflerc@0pointer.net>  Sun, 12 Feb 2006 11:49:48 +0100

syrep (0.7-1) unstable; urgency=high

  * New upstream

 -- Lennart Poettering <mzflerc@0pointer.de>  Fri, 15 Oct 2004 13:10:57 +0200
 
syrep (0.6-1) unstable; urgency=high
  
  * New upstream

 -- Lennart Poettering <mzflerc@0pointer.de>  Thu, 30 Sep 2004 22:26:57 +0200

syrep (0.5-1) unstable; urgency=low

  * Initial Release.

 -- Lennart Poettering <mzflerc@0pointer.de>  Fri, 30 Jul 2004 01:03:45 +0200



src/syrep.c
/* $Id: syrep.c 108 2006-04-26 13:01:19Z lennart $ */

/***
  This file is part of syrep.

  syrep is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  syrep is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
  or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
  License for more details.
  
  You should have received a copy of the GNU General Public License
  along with syrep; if not, write to the Free Software Foundation,
  Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
***/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <inttypes.h>
#include <limits.h>
#include <assert.h>
#include <malloc.h>
#include <string.h>
#include <sys/types.h>
#include <dirent.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>
#include <signal.h>

#include <db.h>
#include <zlib.h>

#include "md5util.h"
#include "md5.h"
#include "context.h"
#include "cache.h"
#include "update.h"
#include "list.h"
#include "diff.h"
#include "util.h"
#include "cmdline.h"
#include "syrep.h"
#include "merge.h"
#include "info.h"
#include "history.h"
#include "dump.h"
#include "extract.h"
#include "makepatch.h"
#include "cleanup.h"
#include "forget.h"

#include "svn-revision.h"

volatile int interrupted = 0;

struct gengetopt_args_info args;

static int do_diff(void) {
    struct syrep_db_context *c1 = NULL, *c2 = NULL;
    char *path1 = NULL, *path2 = NULL;
    DB *ddb = NULL;
    int r = 1;

    if (args.inputs_num != 2) {
        fprintf(stderr, "ERROR: Need exactly two repository snapshots for diff command\n");
        goto finish;
    }

    if (args.local_temp_flag && isdirectory(args.inputs[0]) >= 1) {
        const char *p = get_attached_filename(args.inputs[0], SYREP_TEMPDIR);
        mkdir(p, 0777);
        set_tmp_dir(p);
    }

    if (!(path1 = strdup(get_snapshot_filename(args.inputs[0], SYREP_SNAPSHOTFILENAME))))
        goto finish;
    
    if (!(path2 = strdup(get_snapshot_filename(args.inputs[1], SYREP_SNAPSHOTFILENAME))))
        goto finish;
    
    if (!strcmp(path1, path2)) {
        fprintf(stderr, "ERROR: diff command requires two distinct snapshots as arguments\n");
        goto finish;
    }

    if (!(c1 = db_context_open(path1, 0)))
        goto finish;
    
    if (!(c2 = db_context_open(path2, 0)))
        goto finish;
    
    if (!(ddb = make_diff(c1, c2)))
        goto finish;

    if (list_diff(c1, c2, ddb, isdirectory(args.inputs[0]) > 0 ? args.inputs[0] : NULL, isdirectory(args.inputs[1]) > 0 ? args.inputs[1] : NULL) < 0)
        goto finish;

    r = 0;
    
finish:
    if (c1)
        db_context_free(c1);

    if (c2)
        db_context_free(c2);

    if (ddb)
        ddb->close(ddb, 0);

    if (path1)
        free(path1);

    if (path2)
        free(path2);

    return r;
}

static int do_merge(void) {
    struct syrep_db_context *c1 = NULL, *c2 = NULL;
    char *path1 = NULL, *path2 = NULL;
    int r = 1;

    if (args.inputs_num != 2) {
        fprintf(stderr, "ERROR: Need exactly one repository snapshot and one repository for merge command\n");
        goto finish;
    }

    if (isdirectory(args.inputs[1]) <= 0) {
        fprintf(stderr, "ERROR: %s is not a directory\n", args.inputs[1]);
        goto finish;
    }

    if (args.local_temp_flag) {
        const char *p = get_attached_filename(args.inputs[1], SYREP_TEMPDIR);
        mkdir(p, 0777);
        set_tmp_dir(p);
    }
    
    if (!(path1 = strdup(get_snapshot_filename(args.inputs[0], SYREP_SNAPSHOTFILENAME))))
        goto finish;
    
    if (!(path2 = strdup(get_snapshot_filename(args.inputs[1], SYREP_SNAPSHOTFILENAME))))
        goto finish;
    
    if (!strcmp(path1, path2)) {
        fprintf(stderr, "ERROR: merge command requires two distinct snapshots as arguments\n");
        goto finish;
    }

    if (!(c1 = db_context_open(path1, 0)))
        goto finish;
    
    if (!(c2 = db_context_open(path2, 0)))
        goto finish;

    if (db_context_origin_warn(c2))
        goto finish;
    
    if (merge(c1, c2, args.inputs[1], isdirectory(args.inputs[0]) > 0 ? args.inputs[0] : NULL) < 0)
        goto finish;

    r = 0;
    
finish:
    if (c1)
        db_context_free(c1);

    if (c2)
        db_context_free(c2);

    if (path1)
        free(path1);

    if (path2)
        free(path2);

    return r;
}

static int do_makepatch(void) {
    struct syrep_db_context *c1 = NULL, *c2 = NULL;
    char *path1 = NULL, *path2 = NULL;
    int r = 1;

    if (args.inputs_num != 2) {
        fprintf(stderr, "ERROR: Need exactly one repository and one repository snapshot for makepatch command\n");
        goto finish;
    }

    if (isdirectory(args.inputs[0]) <= 0) {
        fprintf(stderr, "ERROR: %s is not a directory\n", args.inputs[0]);
        goto finish;
    }
    
    if (args.local_temp_flag) {
        const char *p = get_attached_filename(args.inputs[0], SYREP_TEMPDIR);
        mkdir(p, 0777);
        set_tmp_dir(p);
    }
    
    if (!(path1 = strdup(get_snapshot_filename(args.inputs[0], SYREP_SNAPSHOTFILENAME))))
        goto finish;
    
    if (!(path2 = strdup(get_snapshot_filename(args.inputs[1], SYREP_SNAPSHOTFILENAME))))
        goto finish;
    
    if (!strcmp(path1, path2)) {
        fprintf(stderr, "ERROR: makepatch command requires two distinct snapshots as arguments\n");
        goto finish;
    }

    if (!(c1 = db_context_open(path1, 0)))
        goto finish;
    
    if (db_context_origin_warn(c1))
        goto finish;
    
    if (!(c2 = db_context_open(path2, 0)))
        goto finish;

    if (makepatch(c1, c2, args.inputs[0]) < 0)
        goto finish;

    if (!args.output_file_given && isatty(fileno(stdout)))
        fprintf(stderr, "Sorry, I am not going to write the patch data to a tty.\n");
    else if (db_context_save(c1, args.output_file_given ? args.output_file_arg : NULL) < 0)
        goto finish;

    r = 0;
    
finish:
    if (c1)
        db_context_free(c1);

    if (c2)
        db_context_free(c2);

    if (path1)
        free(path1);

    if (path2)
        free(path2);

    return r;
}

static int do_foreach(int (*func) (struct syrep_db_context *c), int m) {
    char *path = NULL;
    int r = 1;
    unsigned i;
    struct syrep_db_context *c = NULL;

    if (args.inputs_num < 1)
        fprintf(stderr, "WARNING: No repository or snapshot to specified!\n");
    
    for (i = 0; i < args.inputs_num; i++) {
        static char saved_cwd[PATH_MAX];

        if (args.local_temp_flag && isdirectory(args.inputs[i]) >= 1) {
            const char *p = get_attached_filename(args.inputs[i], SYREP_TEMPDIR);
            mkdir(p, 0777);
            set_tmp_dir(p);
        }
        
        if (!(path = strdup(get_snapshot_filename(args.inputs[i], SYREP_SNAPSHOTFILENAME))))
            goto finish;
        
        if (!(c = db_context_open(path, 0)))
            goto finish;

        if (args.inputs_num > 1)
            fprintf(stderr, "*** %s ***\n", path);
        
        if (m && args.output_directory_given) {
            if (!getcwd(saved_cwd, sizeof(saved_cwd))) {
                fprintf(stderr, "getcwd(): %s\n", strerror(errno));
                return -1;
            }
            
            mkdir(args.output_directory_arg, 0777);
            
            if (chdir(args.output_directory_arg) < 0) {
                fprintf(stderr, "Failed to chdir() to directory %s: %s\n", args.output_directory_arg, strerror(errno));
                return -1;
            }
        }
    
        if (func(c) < 0)
            goto finish;

        if (m && args.output_directory_given) {
            if (chdir(saved_cwd) < 0) {
                fprintf(stderr, "Failed to chdir() back to directory %s: %s\n", saved_cwd, strerror(errno));
                return -1;
            }
        }

        if (args.inputs_num > 1 && i < args.inputs_num-1)
            fprintf(stderr, "\n");
        
        db_context_free(c);
        c = NULL;
        
        free(path);
        path = NULL;
    }

    r = 0;
    
finish:
    if (c)
        db_context_free(c);

    if (path)
        free(path);

    return r;
}

static int do_update(void) {
    char *path = NULL;
    int r = 1;
    unsigned i;
    struct syrep_db_context *c = NULL;
    struct syrep_md_cache *cache = NULL;

    if (args.inputs_num < 1)
        fprintf(stderr, "WARNING: No repository to update specified!\n");

    if (args.snapshot_given && args.inputs_num != 1) {
        fprintf(stderr, "ERROR: If a snapshot file is specified only a single directory name is accepted!\n");
        goto finish;
    }
    
    for (i = 0; i < args.inputs_num; i++) {
        static char saved_cwd[PATH_MAX];
        
        if (isdirectory(args.inputs[i]) <= 0) {
            fprintf(stderr, "%s is not a directory\n", args.inputs[i]);
            return 1;
        }

        if (args.local_temp_flag) {
            const char *p = get_snapshot_filename(args.inputs[i], SYREP_TEMPDIR);
            mkdir(p, 0777);
            set_tmp_dir(p);
        }
        
        if (!(path = strdup(get_snapshot_filename(args.inputs[i], SYREP_SNAPSHOTFILENAME))))
            goto finish;
        
        if (!(c = db_context_open(path, 1)))
            goto finish;

        if (db_context_origin_warn(c))
            goto finish;

        if (db_context_fix_origin(c) < 0)
            goto finish;

        if (!args.no_cache_flag) {
            const char *p;
            
            if (args.cache_given)
                cache = md_cache_open(args.cache_arg, args.ro_cache_flag);
            else if ((p = get_attached_filename(args.inputs[i], SYREP_MDCACHEFILENAME)))
                cache = md_cache_open(p, args.ro_cache_flag);
        }

        if (!getcwd(saved_cwd, sizeof(saved_cwd))) {
            fprintf(stderr, "getcwd(): %s\n", strerror(errno));
            goto finish;
        }
        
        if (chdir(args.inputs[i]) < 0) {
            fprintf(stderr, "Failed to chdir() to directory %s: %s\n", args.inputs[i], strerror(errno));
            goto finish;
        }
        
        if (update(c, cache) < 0)
            goto finish;

        if (chdir(saved_cwd) < 0) {
            fprintf(stderr, "Failed to chdir() back to directory %s: %s\n", saved_cwd, strerror(errno));
            goto finish;
        }

        if (db_context_save(c, path) < 0)
            goto finish;

        if (!args.no_purge_flag && cache) 
            md_cache_vacuum(cache); 
        
        db_context_free(c);
        c = NULL;

        free(path);
        path = NULL;

        if (cache) {
            md_cache_close(cache);
            cache = NULL;
        }
    }

    r = 0;
        
finish:
    if (c)
        db_context_free(c);
    
    if (path)
        free(path);

    if (cache)
        md_cache_close(cache);

    return r;
}

static int do_cleanup(void) {
    int r = 1;
    unsigned i;

    if (args.inputs_num < 1)
        fprintf(stderr, "WARNING: No repository specified!\n");
    
    for (i = 0; i < args.inputs_num; i++) {

        if (isdirectory(args.inputs[i]) <= 0) {
            fprintf(stderr, "%s is not a directory\n", args.inputs[i]);
            goto finish;
        }

        if (args.inputs_num > 1)
            fprintf(stderr, "*** %s ***\n", args.inputs[i]);
        
        if (cleanup(args.inputs[i]) < 0)
            goto finish;

        if (args.inputs_num > 1 && i < args.inputs_num-1)
            fprintf(stderr, "\n");
    }

    r = 0;
    
finish:

    return r;
}

static int do_forget(void) {
    char *path = NULL;
    int r = 1;
    unsigned i;
    struct syrep_db_context *c = NULL, *target = NULL;

    if (args.inputs_num < 1)
        fprintf(stderr, "WARNING: No repository or snapshot to specified!\n");
    
    for (i = 0; i < args.inputs_num; i++) {
        if (args.local_temp_flag && isdirectory(args.inputs[i]) >= 1) {
            const char *p = get_attached_filename(args.inputs[i], SYREP_TEMPDIR);
            mkdir(p, 0777);
            set_tmp_dir(p);
        }
        
        if (!(path = strdup(get_snapshot_filename(args.inputs[i], SYREP_SNAPSHOTFILENAME))))
            goto finish;
        
        if (!(c = db_context_open(path, 0)))
            goto finish;

        if (!(target = db_context_open(NULL, 1)))
            goto finish;
        
        if (args.inputs_num > 1)
            fprintf(stderr, "*** %s ***\n", path);

        if (forget(c, target) < 0)
            goto finish;

        if (args.inputs_num > 1 && i < args.inputs_num-1)
            fprintf(stderr, "\n");
        
        db_context_free(c);

        if (db_context_save(target, path) < 0)
            goto finish;
        
        db_context_free(target);
        c = NULL;
        target = NULL;
        
        free(path);
        path = NULL;
    }

    r = 0;
    
finish:
    if (c)
        db_context_free(c);

    if (target)
        db_context_free(target);

    if (path)
        free(path);

    return r;
}

static void sigint(int s) {
    interrupted = 1;
}

static void free_args(void) {
    unsigned i;
    
    for (i = 0; i < args.inputs_num; i++)
        free(args.inputs[i]);

    free(args.inputs);
}

static int help(FILE *f, const char *argv0) {

    fprintf(f,
            "%s -- Synchronize File Repositories\n\n"

            "Usage: %s [options...] <command> [arguments...]\n\n"

            "General options:\n"
            "  -v --verbose                         Enable verbose operation\n"
            "  -T --local-temp                      Use temporary directory inside\n"
            "                                       repository\n"
            "  --ignore-origin                      Don't warn if snapshot not local in\n"
            "                                       update, merge, makepatch\n"
            "  -z --compress                        Compress snapshots or patches\n"
            "  -p --progress                        Show progress\n\n"

            "General commands:\n"
            "  -h --help                            Print help and exit\n"
            "  -V --version                         Print version and exit\n\n"

            "Specific commands:\n"
            "  --list SNAPSHOT                      List a repository snapshot\n"
            "    --show-deleted                     Show deleted entries of repository\n"
            "                                       snapshot\n"
            "    --show-by-md                       Show files by message digests\n"
            "    --show-times                       Show first and last seen times\n"
            "    --sort                             Sort chronologically\n\n"

            "  --info SNAPSHOT                      Show information about a repository or\n"
            "                                       snapshot\n\n" 

            "  --history SNAPSHOT                   Show history of a repository or snapshot\n\n"
            
            "  --dump SNAPSHOT                      Show a structure dump of a repository or\n"
            "                                       snapshot\n\n"
            
            "  --update DIRECTORY                   Update (or create) a repository snapshot\n"
            "    -SSTRING --snapshot=STRING         Use a different snapshot file than the\n"
            "                                       one contained in the repository\n"
            "    -CSTRING --cache=STRING            Use a different cache file than the one\n"
            "                                       contained in the repository\n"
            "    --no-cache                         Don't use a message digest cache\n"
            "    --no-purge                         Don't purge obsolete entries from cache\n"
            "                                       after update run\n"
            "    --ro-cache                         Use read only cache\n"
            "    --check-dev                        Store information about the device where\n"
            "                                       a file resides in the MD cache\n"
            "    --include-dotfiles                 Don't ignore dotfiles\n"
            
            "  --diff SNAPSHOT SNAPSHOT             Show difference between two repositories\n"
            "                                       or snapshots\n"
            "    --sizes -s                         Show sizes for files to copy (works only\n"
            "                                       with repositories)\n"
            "    --human-readable -H                Show sizes in human readable from\n"
            
            "  --merge SNAPSHOT DIRECTORY           Merge a snapshot into a repository\n"
            "                                       (perform deletes, renames only)\n"
            "  --merge PATCH DIRECTORY              Merge a patch into a repository\n"
            "  --merge DIRECTORY DIRECTORY          Merge a repository into a repository\n"
            "    -q --question                      Ask a question before each action\n"
            "    -P --prune-empty                   Prune empty directories\n"
            "    --keep-trash                       Don't empty trash\n"
            "    --check-md                         Check message digest of files prior to\n"
            "                                       deletion or replacement\n"
            "    --always-copy                      Always copy instead of hard link\n\n",
            argv0, argv0);

    fprintf(f, 
            "  --makepatch DIRECTORY SNAPSHOT       Make a patch against the specified\n"
            "                                       repository\n"
            "    -oSTRING --output-file=STRING      Write output to specified file instead\n"
            "                                       of STDOUT\n"
            "    --include-all                      Include files in patch which do exist on\n"
            "                                       the other side under a \n"
            "                                       different name\n\n"
            
            "  --extract SNAPSHOT                   Extract the contents of a snapshot or\n"
            "                                       patch\n"
            "    -DSTRING --output-directory=STRING Write output to specified directory\n\n"
            
            "  --cleanup DIRECTORY                  Remove syrep info from repository\n"
            "    -lINT --cleanup-level=INT          1 - just remove temporary data and trash\n"
            "                                           (default)\n"
            "                                       2 - remove MD cache as well\n"
            "                                       3 - remove all syrep data\n\n"
            "  --forget SNAPSHOT                    Repackage snapshot dropping outdated\n"
            "                                       information\n"
            "    --remember DAYS                    Information of how many days should be\n"
            "                                       kept? (defaults to 180)\n");
            

    return 0;
}

static int version(const char *argv0) {
    int major, minor, patch;

    db_version(&major, &minor, &patch);
    
    fprintf(stderr,
            "%s "PACKAGE_VERSION"\n"
            "Compiled with %i Bit off_t.\n"
            "Compiled with zlib %s, linked to zlib %s.\n"
            "Compiled with libdb %i.%i.%i, linked to libdb %i.%i.%i\n"
            "SVN Revision "SVN_REVISION"\n"
#ifdef USE_XATTR
            "Compiled with extended attribute support: yes\n"
#else
            "Compiled with extended attribute support: no\n"
#endif
#ifdef USE_SENDFILE
            "Using sendfile(): yes\n",
#else
            "Using sendfile(): no\n",
#endif
            argv0,
            sizeof(off_t)*8,
            ZLIB_VERSION, zlibVersion(),
            DB_VERSION_MAJOR, DB_VERSION_MINOR, DB_VERSION_PATCH,
            major, minor, patch );

    return 0;
}

int main(int argc, char *argv[]) {
    char *bn;

    if ((bn = strrchr(argv[0], '/')))
        bn++;
    else
        bn = argv[0];

    signal(SIGINT, sigint);
    
    cmdline_parser(argc, argv, &args);
    atexit(free_args);

    if (args.help_given)
        return help(stdout, bn);
    else if (args.version_given)
        return version(bn);
    else if (args.list_flag) 
        return do_foreach(list, 0);
    else if (args.info_flag) 
        return do_foreach(info, 0);
    else if (args.history_flag)
        return do_foreach(history, 0);
    else if (args.diff_flag)
        return do_diff();
    else if (args.update_flag)
        return do_update();
    else if (args.merge_flag)
        return do_merge();
    else if (args.dump_flag)
        return do_foreach(dump, 0);
    else if (args.extract_flag)
        return do_foreach(extract, 1);
    else if (args.makepatch_flag)
        return do_makepatch();
    else if (args.cleanup_flag)
        return do_cleanup();
    else if (args.forget_flag)
        return do_forget();

    help(stderr, bn);
    
    return 1;
}
